### PR TITLE
L2 MultiMarginCriterion

### PR DESCRIPTION
**Pull request by szagoruyko**
This PR adds p=2 in:
```lua
loss(x,y) = forward(x,y) = sum_i(max(0, 1 - (x[y] - x[i]))^p) / x:size(1)
```
Also added test, cuda version is coming.

### CODE REVIEW

File: generic/MultiMarginCriterion.c [hunk 1]
```diff
@@ -54,7 +55,7 @@ static int nn_(MultiMarginCriterion_updateOutput)(lua_State *L)
         continue;
     
       if(z > 0)
-        sum += z;
+        sum += (p==1) ? z : z*z;
```
Comment (soumith) [inline generic/MultiMarginCriterion.c]: this only works for p=1 and p=2.

File: generic/MultiMarginCriterion.c [hunk 2]
```diff
@@ -127,8 +129,9 @@ static int nn_(MultiMarginCriterion_updateGradInput)(lua_State *L)
     
       if(z > 0)
       {
-        gradInput_target -= g;
-        gradInput_data[d] = g;
+        real h = (p == 1) ? g : 2*g*z;
```
Comment (soumith) [inline generic/MultiMarginCriterion.c]: same, only works for p=1 or p=2. if this is by design, assert in the constructor. if not, fix it appropriately.

Comment (szagoruyko) [inline generic/MultiMarginCriterion.c]: added an assert in constructor for p=1 and p=2

**Comment by soumith** [comment]
thanks!

**Pull request merged**